# ĐỒ ÁN CUỐI KÌ
Môn học: **Lập trình khoa học dữ liệu**

## Thông tin thành viên
| STT | Họ tên | MSSV |
|:-------|:------:| :------:|
|  1  |  Nguyễn Gia Hân | 19127134 |
|  2  |  Phạm Minh Tú  | 19127611 |


## I. Import các thư viện cần thiết

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


## II. Thu thập dữ liệu

### Chủ đề của dữ liệu là gì ? Nguồn của dữ liệu

Chủ đề của dữ liệu: Thông tin tuyển dụng về ngành khoa học dữ liệu trên www.glassdoor.com \
Nguồn của dữ liệu: https://www.kaggle.com/rashikrahmanpritom/data-science-job-posting-on-glassdoor?fbclid=IwAR3kOAZB29_Dq5bubOHASAbqnU7n4-a3eSwmwbwM_3rPRqZmVyIK6ntSxZ0

### Tác giả có cho phép bạn sử dụng dữ liệu? Kiểm tra về giấy phép của dữ liệu

Tác giả có cho phép sử dụng tài liệu này. \
Theo license: CC0: Public Domain.

### Tác giả đã thu thập dữ liệu như thế nào?

Dữ liệu này được thu thập bằng cách Web scrapping, Data cleaning & EDA code các tinh tuyển dụng liên quan đến vị trí 'Data Scientist' từ www.glassdoor.com ở USA.

## III. Khám phá dữ liệu

### Đọc dữ liệu từ file csv vào dataframe

In [2]:
df = pd.read_csv('data.csv')
df

Job Title Salary Estimate  \
0    Sr Data Scientist        137-171    
1       Data Scientist        137-171    
2       Data Scientist        137-171    
3       Data Scientist        137-171    
4       Data Scientist        137-171    
..                 ...             ...   
655     Data Scientist        105-167    
656     Data Scientist        105-167    
657     Data Scientist        105-167    
658     Data Scientist        105-167    
659     Data Scientist        105-167    

                                       Job Description  Rating  \
0    Description\r\n\r\nThe Senior Data Scientist i...     3.1   
1    Secure our Nation, Ignite your Future\r\n\r\nJ...     4.2   
2    Overview\r\n\r\n\r\nAnalysis Group is one of t...     3.8   
3    JOB DESCRIPTION:\r\n\r\nDo you have a passion ...     3.5   
4    Data Scientist\r\nAffinity Solutions / Marketi...     2.9   
..                                                 ...     ...   
655  Summary\r\n\r\nWe’re looking for a data scient...     3.6   
656  Job Description\r\nBecome a thought leader wit...     0.0   
657  Join a thriving company that is changing the w...     0.0   
658  100 Remote Opportunity As an AINLP Data Scient...     5.0   
659  Description\r\n\r\nThe Data Scientist will be ...     2.7   

             Company Name           Location            Headquarters  \
0             Healthfirst       New York, NY            New York, NY   
1                 ManTech      Chantilly, VA             Herndon, VA   
2          Analysis Group         Boston, MA              Boston, MA   
3                 INFICON         Newton, MA  Bad Ragaz, Switzerland   
4      Affinity Solutions       New York, NY            New York, NY   
..                    ...                ...                     ...   
655              TRANZACT       Fort Lee, NJ            Fort Lee, NJ   
656                  JKGT  San Francisco, CA                      -1   
657            AccessHope      Irwindale, CA                      -1   
658  ChaTeck Incorporated  San Francisco, CA         Santa Clara, CA   
659         1-800-Flowers       New York, NY         Carle Place, NY   

                        Size        Type of ownership  \
0     1001 to 5000 employees   Nonprofit Organization   
1    5001 to 10000 employees         Company - Public   
2     1001 to 5000 employees  Private Practice / Firm   
3      501 to 1000 employees         Company - Public   
4        51 to 200 employees        Company - Private   
..                       ...                      ...   
655   1001 to 5000 employees        Company - Private   
656                       -1                       -1   
657                       -1                       -1   
658        1 to 50 employees        Company - Private   
659   1001 to 5000 employees         Company - Public   

                                  Industry  ... company_age python  excel  \
0                       Insurance Carriers  ...          27      0      0   
1                   Research & Development  ...          52      0      0   
2                               Consulting  ...          39      1      1   
3    Electrical & Electronic Manufacturing  ...          20      1      1   
4                  Advertising & Marketing  ...          22      1      1   
..                                     ...  ...         ...    ...    ...   
655                Advertising & Marketing  ...          31      1      1   
656                                     -1  ...          -1      0      0   
657                                     -1  ...          -1      1      1   
658                Advertising & Marketing  ...          -1      1      0   
659                              Wholesale  ...          44      1      1   

     hadoop  spark aws  tableau  big_data        job_simp  seniority  
0         0      0   1        0         0  data scientist     senior  
1         1      0   0        0         1  data scientist         na  
2         0      0   1     

### Dữ liệu gồm có bao nhiêu dòng và bao nhiêu cột?

In [3]:
n_rows, n_cols  = df.shape
n_rows, n_cols

(660, 27)

Dữ liệu có 660 dòng, 27 cột

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Theo file mô tả dữ liệu và theo quan sát sơ bộ thì một dòng là thông tin của một bài đăng tuyển dụng nhân sự của công ty về các vị trí công việc liên quan đến 'Data Science'. \
Có vẻ như không có dòng nào có nội dung lệch lạc so với chủ đề của dữ liệu.

### Dữ liệu có các dòng bị lặp không?

In [4]:
duplicated_rows = df[df.duplicated()]
have_duplicated_rows = len(duplicated_rows) != 0
have_duplicated_rows

True

Vậy dữ liệu này có các dòng bị trùng lặp

**Tiền xử lí**  
Loại bỏ các dòng bị trùng lặp

In [5]:
df = df.drop_duplicates()
df

Job Title Salary Estimate  \
0    Sr Data Scientist        137-171    
1       Data Scientist        137-171    
2       Data Scientist        137-171    
3       Data Scientist        137-171    
4       Data Scientist        137-171    
..                 ...             ...   
655     Data Scientist        105-167    
656     Data Scientist        105-167    
657     Data Scientist        105-167    
658     Data Scientist        105-167    
659     Data Scientist        105-167    

                                       Job Description  Rating  \
0    Description\r\n\r\nThe Senior Data Scientist i...     3.1   
1    Secure our Nation, Ignite your Future\r\n\r\nJ...     4.2   
2    Overview\r\n\r\n\r\nAnalysis Group is one of t...     3.8   
3    JOB DESCRIPTION:\r\n\r\nDo you have a passion ...     3.5   
4    Data Scientist\r\nAffinity Solutions / Marketi...     2.9   
..                                                 ...     ...   
655  Summary\r\n\r\nWe’re looking for a data scient...     3.6   
656  Job Description\r\nBecome a thought leader wit...     0.0   
657  Join a thriving company that is changing the w...     0.0   
658  100 Remote Opportunity As an AINLP Data Scient...     5.0   
659  Description\r\n\r\nThe Data Scientist will be ...     2.7   

             Company Name           Location            Headquarters  \
0             Healthfirst       New York, NY            New York, NY   
1                 ManTech      Chantilly, VA             Herndon, VA   
2          Analysis Group         Boston, MA              Boston, MA   
3                 INFICON         Newton, MA  Bad Ragaz, Switzerland   
4      Affinity Solutions       New York, NY            New York, NY   
..                    ...                ...                     ...   
655              TRANZACT       Fort Lee, NJ            Fort Lee, NJ   
656                  JKGT  San Francisco, CA                      -1   
657            AccessHope      Irwindale, CA                      -1   
658  ChaTeck Incorporated  San Francisco, CA         Santa Clara, CA   
659         1-800-Flowers       New York, NY         Carle Place, NY   

                        Size        Type of ownership  \
0     1001 to 5000 employees   Nonprofit Organization   
1    5001 to 10000 employees         Company - Public   
2     1001 to 5000 employees  Private Practice / Firm   
3      501 to 1000 employees         Company - Public   
4        51 to 200 employees        Company - Private   
..                       ...                      ...   
655   1001 to 5000 employees        Company - Private   
656                       -1                       -1   
657                       -1                       -1   
658        1 to 50 employees        Company - Private   
659   1001 to 5000 employees         Company - Public   

                                  Industry  ... company_age python  excel  \
0                       Insurance Carriers  ...          27      0      0   
1                   Research & Development  ...          52      0      0   
2                               Consulting  ...          39      1      1   
3    Electrical & Electronic Manufacturing  ...          20      1      1   
4                  Advertising & Marketing  ...          22      1      1   
..                                     ...  ...         ...    ...    ...   
655                Advertising & Marketing  ...          31      1      1   
656                                     -1  ...          -1      0      0   
657                                     -1  ...          -1      1      1   
658                Advertising & Marketing  ...          -1      1      0   
659                              Wholesale  ...          44      1      1   

     hadoop  spark aws  tableau  big_data        job_simp  seniority  
0         0      0   1        0         0  data scientist     senior  
1         1      0   0        0         1  data scientist         na  
2         0      0   1     

### Mỗi cột có ý nghĩa gì?

Theo file mô tả dữ liệu thì: 
- Job Title: Tên của công việc được tuyển dụng 
- Salary Estimation: Mức lương cho công việc ( tối thiếu - tối đa )
- Job Description: Mô tả đầy đủ về công việc 
- Rating: Đánh giá của mọi người về bài đăng tuyển dụng 
- Company: Tên công ty 
- Location: Vị trí của công ty 
- Headquarter: Vị trí của trụ sở chính 
- Size: Tổng số nhân viên trong công ty đó 
- Type of ownership: Mô tả loại hình công ty, tức là trang trại phi lợi nhuận / nhà nước / tư nhân, v.v. 
- Industry, Sector: Lĩnh vực mà công ty đó làm việc
- Revenue: Tổng doanh thu của công ty
- min_salary, max_salary, avg_salary: Lương tối thiếu, tối đa, trung bình của công việc
- job_state: Nơi ứng viên sẽ làm việc
- same_state: Cùng nơi với trụ sở chính hay không (có hay không)
- company_age: Tuổi đời của công ty
- python,excel,hadoop,spark,aws,tableau,big_data: Kĩ năng yêu cầu ở công việc (có hay không)
- job_simp: Loại chuyên môn
- seniority: Yêu cầu kinh nghiệm làm việc

**Tiền xử lí**  
Ta thấy:   
- Cột minsalary, maxsalary đã bao gồm cột Salary Estimation, và từ minsalary, maxsalary có thể suy ra avgsalary
- Cột Sector có ý nghĩa bao quát hơn cột Industry  
    -> Xóa 4 cột minsalary, maxsalary, avgsalary, Industry   
    => Vậy nên cần tiền xử lí trước khi tiếp tục khám phá dữ liệu  
- Ở trong dữ liệu này, các giá trị thiếu được biểu diễn bằng hai giá trị là  '-1' và 'na' cho nên cần biểu diễn giá trị thiếu thành None 


In [6]:
columns = ['Salary Estimate', 'avg_salary', 'Industry']

df = df.drop(columns, axis=1)

df = df.replace({'-1': None})
df = df.replace({'na': None})
df['company_age'] = df['company_age'].replace({-1: np.nan})
df

Job Title                                    Job Description  \
0    Sr Data Scientist  Description\r\n\r\nThe Senior Data Scientist i...   
1       Data Scientist  Secure our Nation, Ignite your Future\r\n\r\nJ...   
2       Data Scientist  Overview\r\n\r\n\r\nAnalysis Group is one of t...   
3       Data Scientist  JOB DESCRIPTION:\r\n\r\nDo you have a passion ...   
4       Data Scientist  Data Scientist\r\nAffinity Solutions / Marketi...   
..                 ...                                                ...   
655     Data Scientist  Summary\r\n\r\nWe’re looking for a data scient...   
656     Data Scientist  Job Description\r\nBecome a thought leader wit...   
657     Data Scientist  Join a thriving company that is changing the w...   
658     Data Scientist  100 Remote Opportunity As an AINLP Data Scient...   
659     Data Scientist  Description\r\n\r\nThe Data Scientist will be ...   

     Rating          Company Name           Location            Headquarters  \
0       3.1           Healthfirst       New York, NY            New York, NY   
1       4.2               ManTech      Chantilly, VA             Herndon, VA   
2       3.8        Analysis Group         Boston, MA              Boston, MA   
3       3.5               INFICON         Newton, MA  Bad Ragaz, Switzerland   
4       2.9    Affinity Solutions       New York, NY            New York, NY   
..      ...                   ...                ...                     ...   
655     3.6              TRANZACT       Fort Lee, NJ            Fort Lee, NJ   
656     0.0                  JKGT  San Francisco, CA                    None   
657     0.0            AccessHope      Irwindale, CA                    None   
658     5.0  ChaTeck Incorporated  San Francisco, CA         Santa Clara, CA   
659     2.7         1-800-Flowers       New York, NY         Carle Place, NY   

                        Size        Type of ownership             Sector  \
0     1001 to 5000 employees   Nonprofit Organization          Insurance   
1    5001 to 10000 employees         Company - Public  Business Services   
2     1001 to 5000 employees  Private Practice / Firm  Business Services   
3      501 to 1000 employees         Company - Public      Manufacturing   
4        51 to 200 employees        Company - Private  Business Services   
..                       ...                      ...                ...   
655   1001 to 5000 employees        Company - Private  Business Services   
656                     None                     None               None   
657                     None                     None               None   
658        1 to 50 employees        Company - Private  Business Services   
659   1001 to 5000 employees         Company - Public  Business Services   

                        Revenue  ...  company_age  python excel  hadoop  \
0      Unknown / Non-Applicable  ...         27.0       0     0       0   
1        $1 to $2 billion (USD)  ...         52.0       0     0       1   
2    $100 to $500 million (USD)  ...         39.0       1     1       0   
3    $100 to $500 million (USD)  ...         20.0       1     1       0   
4      Unknown / Non-Applicable  ...         22.0       1     1       0   
..                          ...  ...          ...     ...   ...     ...   
655    Unknown / Non-Applicable  ...         31.0       1     1       1   
656                        None  ...          NaN       0     0       0   
657                        None  ...          NaN       1     1       1   
658      $1 to $5 million (USD)  ...          NaN       1     0       1   
659      $1 to $2 billion (USD)  ...         44.0       1     1       0   

     spark  aws  tableau  big_data        job_simp  seniority  
0        0    1        0         0  data scientist     senior  
1        0    0        0         1  data scientist       None  
2        0    1        0         0  data scientist       None  
3        0    1        0         0  data scientist       

### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

In [7]:
df.dtypes

Job Title             object
Job Description       object
Rating               float64
Company Name          object
Location              object
Headquarters          object
Size                  object
Type of ownership     object
Sector                object
Revenue               object
min_salary             int64
max_salary             int64
job_state             object
same_state             int64
company_age          float64
python                 int64
excel                  int64
hadoop                 int64
spark                  int64
aws                    int64
tableau                int64
big_data               int64
job_simp              object
seniority             object
dtype: object

Hiện tại, các cột có vẻ đã có kiểu dữ liệu phù hợp để xử lý tiếp  
Các cột same_state, python, excel, hadoop, spark, aws, tableau, big_data đáng lẽ sẽ có kiểu dữ liệu boolean, tuy nhiên nhóm thấy để dạng int64 vẫn có thể dễ dàng để xử lý, vậy nên nhóm sẽ để nguyên.

### Với mỗi cột có kiểu dữ liệu dạng numeric, các giá trị được phân bố như thế nào?
Phần trăm các giá trị thiếu là bao nhiêu?  
Min? max? Có gì bất thường không?

In [8]:
def missing_ratio (col):
    return col.isna().mean()*100

num_cols = df.select_dtypes(include=np.number)
num_cols_profiles = num_cols.agg([missing_ratio, pd.Series.min, pd.Series.max])
num_cols_profiles

Rating  min_salary  max_salary  same_state  company_age  \
missing_ratio     0.0         0.0         0.0         0.0    16.512346   
min               0.0        31.0        56.0         0.0     1.000000   
max               5.0       212.0       331.0         1.0   239.000000   

               python  excel  hadoop  spark  aws  tableau  big_data  
missing_ratio     0.0    0.0     0.0    0.0  0.0      0.0       0.0  
min               0.0    0.0     0.0    0.0  0.0      0.0       0.0  
max               1.0    1.0     1.0    1.0  1.0      1.0       1.0

Cột Rating đánh giá từ 0 đến 5 nên min, max như vậy là chính xác, không có giá trị thiếu  
Cột min_salary, max_salary không có giá trị thiếu và không có gì bất thường ở min max  
Cột same_stage, python, excel, hadoop, spark, aws, tableau,	big_data lấy 0 là True, 1 là False, vậy nên min max đúng, không có giá trị thiếu  
Cột company_age min, max không có gì lạ, dữ liệu thiếu có thể do công ty không cung cấp tuổi đời, không có gì bất thường  
-> Toàn bộ không có gì bất thường  

### Với mỗi cột có kiểu dữ liệu dạng categorical, các giá trị được phân bố như thế nào? 
Phần trăm các giá trị thiếu là bao nhiêu?  
Có bao nhiêu giá trị khác nhau? Hiển thị một vài  
Có gì bất thường không?

In [9]:
def diff_vals(s):
    return s.dropna().unique()

def num_diff_vals(s):
    return s.dropna().nunique()

cate_cols = df.select_dtypes(exclude=np.number)
cate_cols_profiles= cate_cols.agg([missing_ratio, num_diff_vals, diff_vals])
cate_cols_profiles

Job Title  \
missing_ratio                                                0.0   
num_diff_vals                                                168   
diff_vals      [Sr Data Scientist, Data Scientist, Data Scien...   

                                                 Job Description  \
missing_ratio                                                0.0   
num_diff_vals                                                481   
diff_vals      [Description\r\n\r\nThe Senior Data Scientist ...   

                                                    Company Name  \
missing_ratio                                                0.0   
num_diff_vals                                                425   
diff_vals      [Healthfirst, ManTech, Analysis Group, INFICON...   

                                                        Location  \
missing_ratio                                                0.0   
num_diff_vals                                                202   
diff_vals      [New York, NY, Chantilly, VA, Boston, MA, Newt...   

                                                    Headquarters  \
missing_ratio                                            3.08642   
num_diff_vals                                                226   
diff_vals      [New York, NY, Herndon, VA, Boston, MA, Bad Ra...   

                                                            Size  \
missing_ratio                                           2.469136   
num_diff_vals                                                  8   
diff_vals      [1001 to 5000 employees, 5001 to 10000 employe...   

                                               Type of ownership  \
missing_ratio                                           2.469136   
num_diff_vals                                                 12   
diff_vals      [Nonprofit Organization, Company - Public, Pri...   

                                                          Sector  \
missing_ratio                                           9.259259   
num_diff_vals                                                 22   
diff_vals      [Insurance, Business Services, Manufacturing, ...   

                                                         Revenue  \
missing_ratio                                           2.469136   
num_diff_vals                                                 13   
diff_vals      [Unknown / Non-Applicable, $1 to $2 billion (U...   

                                                       job_state  \
missing_ratio                                                0.0   
num_diff_vals                                                 39   
diff_vals      [ NY,  VA,  MA,  CA,  IL,  MO,  WA,  DC,  TN, ...   

                                                        job_simp     seniority  
missing_ratio                                          10.493827     85.802469  
num_diff_vals                                                  6             2  
diff_vals      [data scientist, analyst, data engineer, mle, ...  [senior, jr]

Cột Job Title, Job Description, job_state không có giá trị thiếu, số giá trị khác nhau ít hơn số dòng vì có nhiều tên job, mô tả job, nơi làm của job giống nhau không có gì lạ  
Cột Company Name không có giá trị thiếu, số giá trị khác nhau ít hơn số dòng vì một công ty có thể đăng nhiều tuyển dụng  
Cột Location không có giá trị thiếu, số giá trị khác nhau ít hơn số dòng vì một vị trí co thể có nhiều công ty  
Cột Headquarters, Size, Type of ownership, Sector, Revenue có giá trị thiếu vì có nhiều công ty không có hoặc không cung cấp các thông tin khi đăng kí trên trang tuyển dụng  
Cột job_simp có giá trị thiếu vì có thể công việc có không nằm trong loại chuyên môn cụ thể nào, và chỉ có 6 giá trị khác nhau vì một chuyên môn có thể có nhiều công việc
Cột seniority có nhiều giá trị thiếu vì có nhiều công việc không yêu cầu cụ thể trình độ  
-> Toàn bộ không có gì bất thường

## IV. Đưa ra các câu hỏi cần trả lời
Câu hỏi là gì?  
Lợi ích của việc tìm ra câu trả lời là gì?

1. Top 10 công việc được tuyển dụng có chỉ số đánh giá cao. Liệu lương có phải là yếu tố ảnh hướng nhất đến kết quả đánh giá đó hay không.  
    ->  
2. Có phải các công ty thuộc quyền sở hữu tư nhân sẽ trả lương cao và đánh giá tốt hơn so với các công ty khác không.  
    ->  
3. Đối với từng chuyên môn trong ngành Data Science, những kĩ năng phổ biến nào trong công việc mà người làm công việc đó cần có.  
    ->  
4. Nơi nào tuyển dụng việc làm về Data Science nhiều nhất. Ở nơi đó, mỗi chuyên môn có công việc tuyển dụng nào được đánh giá cao nhất.  
    ->  

## V. Tiền xử lý + phân tích dữ liệu để trả lời từng câu hỏi

### Câu 1. Top 10 công việc được tuyển dụng có chỉ số đánh giá cao. Liệu lương có phải là yếu tố ảnh hướng nhất đến kết quả đánh giá đó hay không.

### Câu 2. Có phải các công ty thuộc quyền sở hữu tư nhân sẽ trả lương cao và đánh giá tốt hơn so với các công ty khác không ?

### Câu 3. Đối với từng chuyên môn trong ngành Data Science, những kĩ năng phổ biến nào trong công việc mà người làm công việc đó cần có.

### Câu 4: Nơi nào tuyển dụng việc làm về Data Science nhiều nhất. Ở nơi đó, mỗi chuyên môn có công việc tuyển dụng nào được đánh giá cao nhất.

## VI. Reflection
Từng thành viên: 
Bạn đã gặp những khó khăn gì?  
Từng thành viên: Bạn đã học được gì?  
Nhóm của bạn: Nếu có nhiều thời gian hơn, bạn sẽ làm gì?

## VII. References